src/libcore/intrinsics.rs

```diff
@@ -959,59 +959,131 @@ extern "rust-intrinsic" {
     /// value is not necessarily valid to be used to actually access memory.
     pub fn arith_offset<T>(dst: *const T, offset: isize) -> *const T;
 
-    /// Copies `count * size_of<T>` bytes from `src` to `dst`. The source
-    /// and destination may *not* overlap.
+    /// Copies `count * size_of::<T>()` bytes from `src` to `dst`. The source
+    /// and destination must *not* overlap.
     ///
-    /// `copy_nonoverlapping` is semantically equivalent to C's `memcpy`.
+    /// For regions of memory which might overlap, use [`copy`] instead.
+    ///
+    /// `copy_nonoverlapping` is semantically equivalent to C's [`memcpy`].
+    ///
+    /// [`copy`]: ./fn.copy.html
+    /// [`memcpy`]: https://www.gnu.org/software/libc/manual/html_node/Copying-Strings-and-Arrays.html#index-memcpy
     ///
     /// # Safety
     ///
-    /// Beyond requiring that the program must be allowed to access both regions
-    /// of memory, it is Undefined Behavior for source and destination to
-    /// overlap. Care must also be taken with the ownership of `src` and
-    /// `dst`. This method semantically moves the values of `src` into `dst`.
-    /// However it does not drop the contents of `dst`, or prevent the contents
-    /// of `src` from being dropped or used.
+    /// `copy_nonoverlapping` is unsafe because it dereferences a raw pointer.
+    /// The caller must ensure that `src` points to a valid sequence of type
+    /// `T`.
+    ///
+    /// # Undefined Behavior
```

Comment:
I don't see the role of having both `Safety` and `Undefined Behavior` sections. It seems to me that the convention is to have a single section named `Safety`.

```sh
[ lampam @ 21:52:55 ] (master •58 -7 +7) ~/asd/clone/rust
$ rg '# Undefined [bB]ehavior'
src/libcore/mem.rs
532:/// # Undefined behavior

src/liballoc/raw_vec.rs
154:    /// # Undefined Behavior
171:    /// # Undefined Behavior

$ rg '# Safety'
(far more results)
```

Reply:
(granted, I often wish the `# Safety` sections tended to more closely resemble your `# Undefined Behavior` sections, but that is just my opinion)

Reply:
@steveklabnik [mentioned](https://github.com/rust-lang/rust/issues/29371#issuecomment-379234472) that new docs should make use of `Undefined Behavior` (I think upper case is correct). Yeah, the safety section is basically the same for every function. All these functions are trivially unsafe because they dereference raw pointers, and the `Safety` section as I have it doesn't add any new information. Perhaps it could be moved to the module docs and not repeated everywhere?

Reply:
Originally, I felt like these two cases were pretty distinct, but now, re-thinking about it, I'm not so sure, honestly. In some senses, UB is a subset of safety, and it feels like it's worth calling out separately, but maybe not. dunno.

Reply:
I'm confused; I thought UB was the consequence of not following Safety.  Is there a kind of unsafety that doesn't lead to UB?

Reply:
My thought process:

When attempting to divine what belonged under each heading while writing this, I took the point of view that `Safety` referred to the language feature (i.e. the subset of rust you must opt into with the `unsafe` keyword). Some functions (e.g. `Vec::set_len`) are not obviously unsafe from looking at their function signature. The `Safety` section is probably a good place to discuss this.

In discussing why a function is `unsafe`, one must communicate what invariants must be maintained to avoid UB. Most of the standard library uses the `Safety` heading to communicate this. I am now of the opinion that there's never really a reason to have both of these top-level headings (as I do now) since the information is so closely related.

However, for functions whose unsafety is obvious because they take a raw pointer and have names like `read`, `write`, and `copy`, it seemed clearer to name the section discussing its invariants `Undefined Behavior`. `RawVec::from_raw_parts` also uses this convention.

Reply:
@scottmcm "safety" is "memory safety", there are kinds of UB that are not memory safety related.